Add option for cpp_extensions to compile standalone executable

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47865 Enable callgrind collection for C++ snippets
* #47864 Add support for timing C++ snippets.
* #47863 Rework compat bindings.
* **#47862 Add option for cpp_extensions to compile standalone executable**
* #47861 Expose CXX_FLAGS through __config__

Differential Revision: [D25199265](https://our.internmc.facebook.com/intern/diff/D25199265)

Commit:
Add option for cpp_extensions to compile standalone executable

[ghstack-poisoned]

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -23,6 +23,13 @@
 
 IS_WINDOWS = sys.platform == 'win32'
+LIB_EXT = 'pyd' if IS_WINDOWS else 'so'
+EXEC_EXT = 'exe' if IS_WINDOWS else 'o'
```
exec extension is o on linux? Really?

robieta replied:
I wound up just omitting it altogether. (Which I guess is the most UNIX-y thing to do.)

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -1682,7 +1715,9 @@ def object_file_path(source_file: str) -> str:
-    if IS_WINDOWS:
+    if is_standalone:
+        ldflags = extra_ldflags
+    elif IS_WINDOWS:
```
nittiest of nits: it would feel better if it was:

```
if is_standalone:
  ...
else:
  if IS_WINDOWS:
    ...
  else:
    ...
```

as this communicates better that there aren't three cases, there are two cases, and a subdivision in the second case.

robieta replied:
Fair point, though my cyclomatic complexity alarm went off for the double nest. I wound up defining `SHARED_FLAG` since platform is import constant, and then just did `ldflags = ([] if is_standalone else SHARED_FLAG) + extra_ldflags`. A cynic might argue that I'm being too cute / python-y, but hopefully that conveys the intent of "either it's standalone, or slap a `--plays_well_with_others` stamp on it.)

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -1572,7 +1597,12 @@ def _run_ninja_build(build_directory: str, verbose: bool, error_prefix: str) ->
-def _import_module_from_library(module_name, path, is_python_module):
+def _import_module_from_library(module_name, path, is_python_module, is_standalone):
+    if is_standalone:
+        if IS_WINDOWS and TORCH_LIB_PATH not in os.getenv('PATH', ''):
```
Technically this is wrong, because if I have `f'{TORCH_LIB_PATH}/blah/blah` in PATH it will incorrectly detect that the PATH is present when it actually isn't. I also get a bit nervous about forward slash and backward slash confusion here.

robieta replied:
Good point. I added `.split(';')` so now it should actually do the right thing. There shouldn't be any forward-backward slash shenaniganry since I use `os.path.blah` for everything. (Right?)

peterjc123 replied:
It will be strictly safe if you may use `os.path.samefile` and it automatically handles cases and slashes. e.g.
```py
>>> os.path.samefile('C:\\Windows', 'c:/windows')
True
```

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -1572,7 +1597,12 @@ def _run_ninja_build(build_directory: str, verbose: bool, error_prefix: str) ->
 
 
-def _import_module_from_library(module_name, path, is_python_module):
+def _import_module_from_library(module_name, path, is_python_module, is_standalone):
```
I am NOT CONVINCED AT ALL that `is_standalone` should be handled as a boolean flag here. No code is shared between the two function implementations, the return types differ depending on the setting of this value... seems like this should have just been two functions.

robieta replied:
Done.

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -1353,19 +1372,25 @@ def _prepare_ldflags(extra_ldflags, with_cuda, verbose):
+        extra_ldflags.append(f'/LIBPATH:{TORCH_LIB_PATH}')
+        if not is_standalone:
+            extra_ldflags.append('torch_python.lib')
+            extra_ldflags.append(f'/LIBPATH:{python_lib_path}')
```
By the way, one could conceivable want a standalone executable that still has the Python interpreter. This use case would be when the top level runner is a C++ program, but then calls into the Python runtime. Pretty obscure, but it's a valid configuration. (No action needed.)

ezyang commented on test/test_utils.py:
```diff
@@ -642,5 +645,56 @@ def test_assert_true(self):
+                    stdout=subprocess.PIPE,
+                )
+                self.assertEqual(r.returncode, 0)
+                self.assertEqual(
```
expect test?

robieta replied:
In this case no, because:
1) `expecttest` doesn't handle the loop properly. (Tries to double update.)
2) The replacement is completely dedented.

I tried it (factoring the run bits so there were two literal tests) and it was just ugly and awkward. If there were more test cases or the output was less trivial, yeah I'd convert it to an expect test.

ezyang reviewed (approved):
just nits

Commit:
Update on "Add option for cpp_extensions to compile standalone executable"




[ghstack-poisoned]

Commit:
Update on "Add option for cpp_extensions to compile standalone executable"




[ghstack-poisoned]

Commit:
Update on "Add option for cpp_extensions to compile standalone executable"




[ghstack-poisoned]

Commit:
Update on "Add option for cpp_extensions to compile standalone executable"




[ghstack-poisoned]

Commit:
Update on "Add option for cpp_extensions to compile standalone executable"


Differential Revision: [D25199265](https://our.internmc.facebook.com/intern/diff/D25199265)

[ghstack-poisoned]

Commit:
Update on "Add option for cpp_extensions to compile standalone executable"


Differential Revision: [D25199265](https://our.internmc.facebook.com/intern/diff/D25199265)

[ghstack-poisoned]

Commit:
Update on "Add option for cpp_extensions to compile standalone executable"


Differential Revision: [D25199265](https://our.internmc.facebook.com/intern/diff/D25199265)

[ghstack-poisoned]

Commit:
Update on "Add option for cpp_extensions to compile standalone executable"


Differential Revision: [D25199265](https://our.internmc.facebook.com/intern/diff/D25199265)

[ghstack-poisoned]
